core/trino-main/src/main/java/io/trino/sql/planner/plan/TableWriterNode.java

```diff
@@ -309,27 +340,42 @@ public String toString()
         {
             return handle.toString();
         }
+
+        @Override
+        public boolean supportsReportingWrittenBytes(Metadata metadata, Session session)
+        {
+            return metadata.supportsReportingWrittenBytes(session, handle);
+        }
     }
 
     public static class InsertTarget
             extends WriterTarget
     {
-        private final InsertTableHandle handle;
+        private final InsertTableHandle insertTableHandle;
+        private final TableHandle tableHandle;
         private final SchemaTableName schemaTableName;
 
         @JsonCreator
         public InsertTarget(
-                @JsonProperty("handle") InsertTableHandle handle,
-                @JsonProperty("schemaTableName") SchemaTableName schemaTableName)
+                @JsonProperty("handle") InsertTableHandle insertTableHandle,
+                @JsonProperty("schemaTableName") SchemaTableName schemaTableName,
+                @JsonProperty("tableHandle") TableHandle tableHandle)
         {
-            this.handle = requireNonNull(handle, "handle is null");
+            this.insertTableHandle = requireNonNull(insertTableHandle, "insertTableHandle is null");
             this.schemaTableName = requireNonNull(schemaTableName, "schemaTableName is null");
+            this.tableHandle = requireNonNull(tableHandle, "tableHandle is null");
```

Comment:
ditto here. Since it is only needed for validation it is not necessary to serialize the `tableHandle` (in addition to `insertTableHandle`). This can be replaced by a boolean.